raft/log_entry.hpp

```diff
@@ -0,0 +1,58 @@
+// Copyright (C) 2018 Bluzelle
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License, version 3,
+// as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#pragma once
+#include <ostream>
+#include <boost/beast/core/detail/base64.hpp>
+
+namespace {
+
+}
+
+
+namespace bzn {
+    struct log_entry {
+        friend std::ostream& operator << (std::ostream& out, const log_entry& obj)
+        {
+            std::string msg_string = obj.json_to_string(obj.msg);
+            out << obj.log_index << " " << obj.term << " " << boost::beast::detail::base64_encode(msg_string) << "\n";
+            return out;
+        }
+
+
+        friend std::istream&operator >>(std::istream &in,  log_entry& obj)
+        {
+            std::string msg_string;
+            in >> obj.log_index;
+            in >> obj.term;
+            in >> msg_string;
+
+            Json::Reader reader;
+            reader.parse(boost::beast::detail::base64_decode(msg_string), obj.msg);
+            return in;
+        }
+
+
+        inline std::string
+        json_to_string(const bzn::message& msg) const
```

Comment:
Could you not just use toStyledString instead of the writer?

Reply:
toStyledString adds whitespace I'd rather not have in the log. I'd like to keep json_to_string for now, and hopefully remove it when we go to protobuf.

Reply:
Won't matter anyway... the payload is going to come from the client soon base64 encoded.